File: example/1d-linear-convection/weno3/python-v1/02f/src/infrastructure/boundary/neumann.py
# src/boundary/neumann.py
from .base import BoundaryCondition
from core.registry import register_component

@register_component('boundary', 'neumann')
class NeumannBoundary(BoundaryCondition):
    """Neumann（零梯度）边界条件（如出口无梯度）"""
    def apply(self, u):
        nghosts = self.domain.nghosts
        ist = self.domain.ist
        ied = self.domain.ied
        
        # 左边界零梯度
        for ig in range(nghosts):
            u[ist - 1 - ig] = u[ist + ig]
        
        # 右边界零梯度
        for ig in range(nghosts):
            u[ied + ig] = u[ied - 1 - ig]
        
        # 调试信息
        if hasattr(self.config, 'debug') and self.config.debug:
            print(f"  应用Neumann边界: 零梯度")

File: example/1d-linear-convection/weno3/fortran/registry/03f/src/boundary/neumann.f90
! src/boundary/neumann.f90
module neumann_boundary_module
    use base_modules, only: wp, ip
    use boundary_base_module, only: boundary_condition
    implicit none
    private
    
    type, extends(boundary_condition), public :: neumann_boundary
    contains
        procedure :: apply => neumann_apply
    end type
    
    interface neumann_boundary
        module procedure create_neumann_boundary
    end interface
    
contains

    type(neumann_boundary) function create_neumann_boundary() result(this)
        this%name = "neumann"
    end function
    
    subroutine neumann_apply(this, u, nghosts, ist, ied)
        class(neumann_boundary), intent(in) :: this
        real(wp), intent(inout) :: u(:)
        integer(ip), intent(in) :: nghosts, ist, ied
        
        integer :: i
        
        ! 左边界零梯度：u[ist-1-i] = u[ist+i]
        do i = 0, nghosts-1
            u(ist-1-i) = u(ist+i)
        end do
        
        ! 右边界零梯度：u[ied+i] = u[ied-1-i]
        do i = 0, nghosts-1
            u(ied+i) = u(ied-1-i)
        end do
    end subroutine
    
end module neumann_boundary_module